pkg/cmd/subcommand/logs.go

```diff
@@ -111,31 +114,114 @@ func (l *Logs) preRun(cf *genericclioptions.ConfigFlags, args []string) error {
 }
 
 func (l *Logs) run() error {
+	var buidStageDone bool
 	ctx := context.Background()
 	f, err := l.functionClient.CoreV1beta1().Functions(l.namespace).Get(ctx, l.functionName, metav1.GetOptions{})
 	if err != nil {
 		return err
 	}
-	if f.Status.Serving == nil {
-		return nil
+
+	// Serving stage
+serving:
+	if f.Status.Serving != nil {
+		serving := f.Status.Serving.ResourceRef
+
+		err := l.logsForPods(ctx, metav1.ListOptions{LabelSelector: fmt.Sprintf("openfunction.io/serving=%s", serving)})
+		if err != nil {
+			return err
+		}
+	}
+
+	// Build stage
+	if f.Status.Build != nil {
+		if buidStageDone {
+			return nil
+		}
+		buidStageDone = true
+		builderRef := f.Status.Build.ResourceRef
+
+		// get openfunction builder ref
+		builder, err := l.functionClient.CoreV1beta1().Builders(f.Namespace).Get(ctx, builderRef, metav1.GetOptions{})
+		if err != nil {
+			statusError, ok := err.(*k8serrors.StatusError)
+			// builder has been cleaned up
+			if ok && statusError.Status().Code == 404 {
+				goto serving
+			} else {
+				return err
+			}
+		}
+
+		// get shipwright builder-buildrun
+		swbuildrunRef := builder.Status.ResourceRef["shipwright.io/buildRun"]
+		swBuildRun, err := l.swClient.ShipwrightV1alpha1().BuildRuns(f.Namespace).Get(ctx, swbuildrunRef, metav1.GetOptions{})
+		if err != nil {
+			statusError, ok := err.(*k8serrors.StatusError)
+			// buildrun has been cleaned up
+			if ok && statusError.Status().Code == 404 {
+				goto serving
```

Comment:
The `goto serving` logic here is to fetch serving logs continually after the build phase ends.

But I think we can make it simpler by removing the 'goto serving' logic:
- If there is builder CR exists, just continually fetch build logs by adding `-f` option. The builder CR will not be deleted if the build fails and will be deleted when the build succeeds.
- If there is no builder CR, and there is serving CR. This means the build is successful, so we only need to fetch serving logs
- If there is no builder CR and no serving CR either, means this is in between the build and serving phase, and we needn't fetch any logs. And we needn't wait for the serving pod to be started either. The possibility of this scenario is small.

In short, we only need to fetch logs from the currently existing pod's function container by default.
No need to wait for the serving pod to up and running.

Reply:
Yes, I totally agree with you! I thought too much before.

Thanks for your review!